// ssr-common-utils/cjs/common/findRoute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findRoute = void 0;
// @ts-nocheck
const path_to_regexp_1 = require("path-to-regexp");
const cache = {};
const cacheLimit = 10000;
let cacheCount = 0;
function compilePath(path, options) {
    const cacheKey = `${options.end}${options.strict}${options.sensitive}`;
    const pathCache = cache[cacheKey] || (cache[cacheKey] = {});
    if (pathCache[path])
        return pathCache[path];
    const keys = [];
    const regexp = (0, path_to_regexp_1.pathToRegexp)(path, keys, options);
    const result = { regexp, keys };
    if (cacheCount < cacheLimit) {
        pathCache[path] = result;
        cacheCount++;
    }
    return result;
}
function matchPath(pathname, options = {}) {
    if (typeof options === 'string' || Array.isArray(options)) {
        options = { path: options };
    }
    const { path, exact = false, strict = false, sensitive = false, childPath } = options;
    const paths = [].concat(childPath || path);
    return paths.reduce((matched, path) => {
        if (!path && path !== '')
            return null;
        if (matched)
            return matched;
        const { regexp, keys } = compilePath(path, {
            end: exact,
            strict,
            sensitive
        });
        const match = regexp.exec(pathname);
        if (!match)
            return null;
        const [url, ...values] = match;
        const isExact = pathname === url;
        if (exact && !isExact)
            return null;
        return {
            path,
            url: path === '/' && url === '' ? '/' : url,
            isExact,
            params: keys.reduce((memo, key, index) => {
                memo[key.name] = values[index];
                return memo;
            }, {})
        };
    }, null);
}
function findRoute(Routes, path) {
    // 根据请求的path来匹配到对应的Component
    const p = path.includes('?') ? path.split('?')[0] : path;
    const route = Routes.find(route => {
        var _a;
        return route.children ? findRoute(route.children, p) : (_a = matchPath(p, route)) === null || _a === void 0 ? void 0 : _a.isExact;
    });
    return route !== null && route !== void 0 ? route : {};
}
exports.findRoute = findRoute;
