// ssr-common-utils/cjs/server/server-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCtx = exports.useApp = exports.useStoreContext = exports.usePinia = exports.useStore = exports.appLocalStoreageWrapper = exports.localStorageWrapper = void 0;
const async_hooks_1 = require("async_hooks");
const localStorage = new async_hooks_1.AsyncLocalStorage();
exports.localStorageWrapper = {
    run: async function (store, callback) {
        const res = await localStorage.run(store, async function () {
            const res = await callback();
            return res;
        });
        return res;
    },
    getStore: () => localStorage.getStore()
};
const appLocalStoreage = new async_hooks_1.AsyncLocalStorage();
exports.appLocalStoreageWrapper = {
    run: async function (store, callback) {
        const res = await appLocalStoreage.run(store, async function () {
            const res = await callback();
            return res;
        });
        return res;
    },
    getStore: () => appLocalStoreage.getStore()
};
const useStore = () => { var _a; return (_a = localStorage.getStore()) === null || _a === void 0 ? void 0 : _a.store; };
exports.useStore = useStore;
const usePinia = () => { var _a; return (_a = localStorage.getStore()) === null || _a === void 0 ? void 0 : _a.pinia; };
exports.usePinia = usePinia;
const useStoreContext = () => {
    var _a;
    return (_a = exports.localStorageWrapper.getStore()) === null || _a === void 0 ? void 0 : _a.context;
};
exports.useStoreContext = useStoreContext;
const useApp = () => { var _a; return (_a = appLocalStoreage.getStore()) === null || _a === void 0 ? void 0 : _a.app; };
exports.useApp = useApp;
const useCtx = () => { var _a; return (_a = localStorage.getStore()) === null || _a === void 0 ? void 0 : _a.context; };
exports.useCtx = useCtx;
