// ssr-plugin-vue3/cjs/tools/vite.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16;
Object.defineProperty(exports, "__esModule", { value: true });
exports.clientConfig = exports.serverConfig = exports.viteBuildServer = exports.viteBuildClient = exports.viteStart = exports.viteBuild = void 0;
const vite_1 = require("vite");
const ssr_common_utils_1 = require("ssr-common-utils");
const plugin_vue_1 = require("@vitejs/plugin-vue");
const plugin_vue_jsx_1 = require("@vitejs/plugin-vue-jsx");
const plugin_babel_1 = require("@rollup/plugin-babel");
const ssr_vite_plugin_style_import_1 = require("ssr-vite-plugin-style-import");
const { getOutput, vue3ServerEntry, vue3ClientEntry, viteConfig, supportOptinalChaining, isDev, define, optimize } = (0, ssr_common_utils_1.loadConfig)();
const { clientOutPut, serverOutPut } = getOutput();
const styleImportConfig = {
    include: ['**/*.vue', '**/*.ts', '**/*.js', '**/*.tsx', '**/*.jsx', /chunkName/],
    resolves: [
        (0, ssr_vite_plugin_style_import_1.AndDesignVueResolve)(),
        (0, ssr_vite_plugin_style_import_1.VantResolve)(),
        (0, ssr_vite_plugin_style_import_1.ElementPlusResolve)(),
        (0, ssr_vite_plugin_style_import_1.NutuiResolve)(),
        (0, ssr_vite_plugin_style_import_1.AntdResolve)()
    ]
};
const serverPlugins = [
    (0, plugin_vue_1.default)((_b = (_a = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _a === void 0 ? void 0 : _a.server) === null || _b === void 0 ? void 0 : _b.defaultPluginOptions),
    (0, plugin_vue_jsx_1.default)(),
    (_d = (_c = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _c === void 0 ? void 0 : _c.common) === null || _d === void 0 ? void 0 : _d.extraPlugin,
    (_f = (_e = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _e === void 0 ? void 0 : _e.server) === null || _f === void 0 ? void 0 : _f.extraPlugin,
    (0, ssr_vite_plugin_style_import_1.createStyleImportPlugin)(styleImportConfig),
    !supportOptinalChaining && (0, plugin_babel_1.default)({
        babelHelpers: 'bundled',
        plugins: [
            '@babel/plugin-proposal-optional-chaining',
            '@babel/plugin-proposal-nullish-coalescing-operator'
        ],
        exclude: /node_modules|\.(css|less|sass)/,
        extensions: ['.vue', '.ts', '.tsx', '.js']
    })
].filter(Boolean);
const serverConfig = {
    ...(0, ssr_common_utils_1.commonConfig)(),
    ...(_g = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().server) === null || _g === void 0 ? void 0 : _g.otherConfig,
    plugins: (_l = (_k = (_j = (_h = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _h === void 0 ? void 0 : _h.server) === null || _j === void 0 ? void 0 : _j.processPlugin) === null || _k === void 0 ? void 0 : _k.call(_j, serverPlugins)) !== null && _l !== void 0 ? _l : serverPlugins,
    optimizeDeps: {
        ...(_o = (_m = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().server) === null || _m === void 0 ? void 0 : _m.otherConfig) === null || _o === void 0 ? void 0 : _o.optimizeDeps,
        esbuildOptions: {
            ...(_r = (_q = (_p = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().server) === null || _p === void 0 ? void 0 : _p.otherConfig) === null || _q === void 0 ? void 0 : _q.optimizeDeps) === null || _r === void 0 ? void 0 : _r.esbuildOptions,
            // @ts-expect-error
            bundle: isDev
        }
    },
    build: {
        ...(_t = (_s = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().server) === null || _s === void 0 ? void 0 : _s.otherConfig) === null || _t === void 0 ? void 0 : _t.build,
        ssr: vue3ServerEntry,
        outDir: serverOutPut,
        rollupOptions: {
            ...(_w = (_v = (_u = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().server) === null || _u === void 0 ? void 0 : _u.otherConfig) === null || _v === void 0 ? void 0 : _v.build) === null || _w === void 0 ? void 0 : _w.rollupOptions,
            input: isDev ? vue3ClientEntry : vue3ServerEntry,
            output: {
                entryFileNames: 'Page.server.js',
                assetFileNames: (0, ssr_common_utils_1.rollupOutputOptions)().assetFileNames
            }
        }
    },
    define: {
        ...(0, ssr_common_utils_1.getDefineEnv)(),
        ...(_y = (_x = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().server) === null || _x === void 0 ? void 0 : _x.otherConfig) === null || _y === void 0 ? void 0 : _y.define,
        __isBrowser__: false,
        ...define === null || define === void 0 ? void 0 : define.base,
        ...define === null || define === void 0 ? void 0 : define.server
    }
};
exports.serverConfig = serverConfig;
const clientPlugins = [
    (0, plugin_vue_1.default)((_0 = (_z = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _z === void 0 ? void 0 : _z.client) === null || _0 === void 0 ? void 0 : _0.defaultPluginOptions),
    (0, plugin_vue_jsx_1.default)(),
    (_2 = (_1 = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _1 === void 0 ? void 0 : _1.common) === null || _2 === void 0 ? void 0 : _2.extraPlugin,
    (_4 = (_3 = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _3 === void 0 ? void 0 : _3.client) === null || _4 === void 0 ? void 0 : _4.extraPlugin,
    (0, ssr_vite_plugin_style_import_1.createStyleImportPlugin)(styleImportConfig)
].filter(Boolean);
const clientConfig = {
    ...(0, ssr_common_utils_1.commonConfig)(),
    ...(_5 = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().client) === null || _5 === void 0 ? void 0 : _5.otherConfig,
    base: isDev ? '/' : (0, ssr_common_utils_1.getOutputPublicPath)(),
    plugins: (_9 = (_8 = (_7 = (_6 = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig()) === null || _6 === void 0 ? void 0 : _6.client) === null || _7 === void 0 ? void 0 : _7.processPlugin) === null || _8 === void 0 ? void 0 : _8.call(_7, clientPlugins)) !== null && _9 !== void 0 ? _9 : clientPlugins,
    build: {
        ...(_11 = (_10 = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().client) === null || _10 === void 0 ? void 0 : _10.otherConfig) === null || _11 === void 0 ? void 0 : _11.build,
        ...(optimize ? { write: false } : {}),
        ssrManifest: true,
        outDir: clientOutPut,
        rollupOptions: {
            ...(_14 = (_13 = (_12 = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().client) === null || _12 === void 0 ? void 0 : _12.otherConfig) === null || _13 === void 0 ? void 0 : _13.build) === null || _14 === void 0 ? void 0 : _14.rollupOptions,
            input: vue3ClientEntry,
            output: (0, ssr_common_utils_1.rollupOutputOptions)(),
            plugins: [(0, ssr_common_utils_1.chunkNamePlugin)(), (0, ssr_common_utils_1.asyncOptimizeChunkPlugin)(), (0, ssr_common_utils_1.manifestPlugin)(),
                ...(0, ssr_common_utils_1.getBabelOptions)({
                    babel: plugin_babel_1.default
                })]
        }
    },
    define: {
        ...(0, ssr_common_utils_1.getDefineEnv)(),
        ...(_16 = (_15 = viteConfig === null || viteConfig === void 0 ? void 0 : viteConfig().client) === null || _15 === void 0 ? void 0 : _15.otherConfig) === null || _16 === void 0 ? void 0 : _16.define,
        __isBrowser__: true,
        ...define === null || define === void 0 ? void 0 : define.base,
        ...define === null || define === void 0 ? void 0 : define.client
    }
};
exports.clientConfig = clientConfig;
const viteStart = async () => {
    //
};
exports.viteStart = viteStart;
const viteBuild = async () => {
    await (0, vite_1.build)({ ...clientConfig, mode: 'production' });
    await (0, vite_1.build)({ ...serverConfig, mode: 'production' });
};
exports.viteBuild = viteBuild;
const viteBuildClient = async () => {
    await (0, vite_1.build)({ ...clientConfig, mode: 'production' }).catch(_ => { });
};
exports.viteBuildClient = viteBuildClient;
const viteBuildServer = async () => {
    await (0, vite_1.build)({ ...serverConfig, mode: 'production' });
};
exports.viteBuildServer = viteBuildServer;
