// ssr-plugin-vue3/cjs/entry/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createStore = exports.createRouter = exports.getVNode = exports.getInlineCssVNode = void 0;
const vue_1 = require("vue");
const vue_router_1 = require("vue-router");
const vuex_1 = require("vuex");
const ssr_deepclone_1 = require("ssr-deepclone");
const combine_router_1 = require("./combine-router");
const { store, FeRoutes } = combine_router_1.Routes;
function createRouter(options = {}) {
    var _a;
    const base = (_a = options.base) !== null && _a !== void 0 ? _a : '/';
    const { hashRouter } = options;
    return (0, vue_router_1.createRouter)({
        history: __isBrowser__ ? (hashRouter ? (0, vue_router_1.createWebHashHistory)(base) : (0, vue_router_1.createWebHistory)(base)) : (0, vue_router_1.createMemoryHistory)(),
        routes: FeRoutes
    });
}
exports.createRouter = createRouter;
function createStore() {
    return (0, vuex_1.createStore)((0, ssr_deepclone_1.deepClone)(store));
}
exports.createStore = createStore;
const getInlineCssVNode = (arr) => arr.map(item => (0, vue_1.h)('style', {
    innerHTML: item
}));
exports.getInlineCssVNode = getInlineCssVNode;
const getVNode = (arr) => arr.map(item => {
    var _a;
    return (0, vue_1.h)((_a = item.tagName) !== null && _a !== void 0 ? _a : 'script', Object.assign({}, item.describe, {
        innerHTML: item.content
    }));
});
exports.getVNode = getVNode;
