// ssr-common-utils/esm/client/client-store.js
const storeCache = {
    val: {
        store: {}
    },
    set: function (store) {
        this.val.store = store;
    },
    get: function () {
        return this.val.store;
    }
};
const piniaCache = {
    val: {
        pinia: {}
    },
    set: function (pinia) {
        this.val.pinia = pinia;
    },
    get: function () {
        return this.val.pinia;
    }
};
const appCache = {
    val: {
        app: {}
    },
    set: function (app) {
        this.val.app = app;
    },
    get: function () {
        return this.val.app;
    }
};
const contextCache = {
    val: {
        obj: {}
    },
    set: function (app) {
        this.val.obj = app;
    },
    get: function () {
        return this.val.obj;
    }
};
export const setStore = (store) => {
    storeCache.set(store);
};
export const useStore = () => {
    return storeCache.get();
};
export const setPinia = (pinia) => {
    piniaCache.set(pinia);
};
export const usePinia = () => {
    return piniaCache.get();
};
export const useCtx = () => {
    console.warn('useCtx can only be used on the server side');
    return {};
};
export const setApp = (app) => {
    appCache.set(app);
};
export const useApp = () => {
    return appCache.get();
};
export const setStoreContext = (context) => {
    contextCache.set(context);
};
export const useStoreContext = () => {
    return contextCache.get();
};
